File: Striver/Day 1 : Array/Stock buy and sell(easy).cpp
//Question: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

//Solution:
class Solution {
public:     //Though this is linear but there is a better approach using DP
    int maxProfit(vector<int>& prices) {
        int max_profit = 0;
        int mini = INT_MAX;
        int profit;
        for (int i=0; i<prices.size(); i++){
            if (prices[i] < mini){
                mini = prices[i];       //keep track of the minimum element on the left of curr element
            }
            profit = prices[i] - mini;
            if (profit > max_profit){
                max_profit = profit;      //try to sell at every point and check the profit
            }
        }
        return max_profit;
    }
};

File: Striver/Day 1 : Array/Stock buy and sell(easy).java

class Solution {
    public int maxProfit(int[] prices) {
        
        
        int min = prices[0];
        int profit = 0;
        int ans = 0;
        
        for(int val : prices)
        {
            min = Math.min(min , val);
            
            profit = val - min;
            
            if(profit > 0)
                ans = Math.max(ans , profit);
                
        }
        
        return ans;
        
    }
}
